keps/sig-scheduling/20190221-even-pods-spreading.md

```diff
@@ -0,0 +1,423 @@
+)
+
+type SpreadingConstraint struct {
+    // MaxSkew describes the degree of imbalance of pods spreading.
+    // It's the max difference between the number of matching pods in any two
+    // topology domains of a given topology type.
+    // For example, in a 3-zone cluster, currently pods with the same labelSelector
+    // are spread as 1/1/0:
+    // - if MaxSkew is 1, incoming pod can only be scheduled to zone3 to become 1/1/1;
+    // schedule it onto zone1(zone2) will make the ActualSkew(2) violates MaxSkew(1)
+    // - if MaxSkew is 2, incoming pod can be scheduled to any zone.
+    // Default value is 1 and 0 is not allowed.
+    MaxSkew int32
+    // TopologyKey is the key such that we consider each value as a "bucket";
+    // we try to put balanced number of pods into each bucket.
+    TopologyKey string
+    // WhenUnsatisfiable indicates how to deal with a pod if it doesn't satisfy
+    // the spreading constraint.
+    // - DoNotSchedule (default) tells the scheduler not to schedule it
+    // - ScheduleAnyway tells the scheduler to still schedule it
+    // Note: it's considered as "Unsatisfiable" only when ActualySkew on all nodes exceeds "MaxSkew".
+    WhenUnsatisfiable ScheduleDecision
+    // Label selector for pods. This's enforced by scheduler to check which pods
+    // should be recognized as a group to satisfy the spreading constraint.
+    Selector *metav1.LabelSelector
+    // namespaces specifies which namespaces the labelSelector applies to (matches against);
+    // null or empty list means "this pod's namespace"
+    // +optional
+    Namespaces []string
+}
+```
+
+### MaxSkew
+
+`MaxSkew` is the core of this KEP, so the exact semantics are clarified as below:
+
+- how Skew is calculated and enforced
+
+Suppose we have a 3-zone cluster, currently pods with the same labelSelector are
+spread as 1/1/0. Internally we compute an "ActualSkew" for each topology
+domain representing "mathing pods in this topology domain" minus "minimum
+matching pods in any topology domain", so for this 1/1/0 cluster, the ActualSkew
+for each zone is 1(1-0)/1(1-0)/0(0-0). (If the spreading is 3/2/1, the
+ActualSkew for each zone will be 2(3-1)/1(2-1)/0(1-1))
+
+The internal computation logic would be to find nodes satifiying "ActualSkew <=
+MaxSkew". Let's go back to the 1/1/0 example:
+
+If MaxSkew is 1, incoming pod can only be scheduled to zone3 to become 1/1/1;
+because schedule it onto zone1(zone2) will make the ActualSkew(2) violates
+MaxSkew(1).
+
+If MaxSkew is 2, incoming pod can be scheduled to any zone.
+
+Things are a little different if `pod.spec.affinity.nodeAffinity` is defined. If
+NodeAffinity is defined, we perfer to assume the user knows the exact topologies
+to deploy to; for example if it's specified as "zone in [zone1, zone2]", and
+current pods are spread as 1/1/0. Suppose MaxSkew is 1, incoming pod can be
+scheduled onto zone1 or zone2 as zone3 has been filtered out (by NodeAffinity).
```

Comment:
@bsalamat Please confirm this is what we want:

> I think we can drop the NodeSelector from EvenSpreading and let the predicate process an external NodeAffinity if there is one in the pod spec.

Reply:
if we plan to deprecate the old affinity fields, we should not rely on its subfields for this feature

Reply:
I think we should clearly say that, when NodeAffinity or NodeSelector features are used, Spreading is applied to nodes that pass those filters. For example, if NodeAffinity chooses zone1 and zone2 and there are 10 zones in the cluster, pods are spread in zone1 and zone2 only and MaxSkew is enforced only on these two zones.

Reply:
I think we should clearly say that, when NodeAffinity or NodeSelector features are used, Spreading is applied to nodes that pass those filters. For example, if NodeAffinity chooses zone1 and zone2 and there are 10 zones in the cluster, pods are spread in zone1 and zone2 only and MaxSkew is enforced only on these two zones.

Reply:
I think we should clearly say that, when NodeAffinity or NodeSelector features are used, Spreading is applied to nodes that pass those filters. For example, if NodeAffinity chooses zone1 and zone2 and there are 10 zones in the cluster, pods are spread in zone1 and zone2 only and MaxSkew is enforced only on these two zones.

Reply:
I think we should clearly say that, when NodeAffinity or NodeSelector features are used, Spreading is applied to nodes that pass those filters. For example, if NodeAffinity chooses zone1 and zone2 and there are 10 zones in the cluster, pods are spread in zone1 and zone2 only and MaxSkew is enforced only on these two zones.

Reply:
Hmm... looks like github has some issues today

Reply:
> if we plan to deprecate the old affinity fields, we should not rely on its subfields for this feature

Oh absolutely we don't deprecate NodeAffinity.

Reply:
Updated the wording based on Bobby's comments.